### PR TITLE
[proof-new] Registering proof checkers uniformly from the SMT solver

### PR DESCRIPTION
**Pull request by HanielB**
Each theory has its own proof checker, responsible for checking the rules pertaining to that theory. The main proof checker uses these specialized checkers. Previously the main proof checker (of the proof node manager used across the SMT solver) was connected to these theory proof checkers during initialization of the theory. This commit adds an interface to the theories for retrieving its proof checker (analogous to how one retrieves the rewriter of a theory) which is used by a new method in the theory engine to register a theory proof checker to a given proof checker according to a theory id.

This is in preparation for the new unsat cores based on proofs.

### CODE REVIEW

File: src/smt/smt_solver.cpp [hunk 1]
```diff
@@ -66,6 +66,16 @@ void SmtSolver::finishInit(const LogicInfo& logicInfo)
     theory::TheoryConstructor::addTheory(d_theoryEngine.get(), id);
   }
 
+  // Add the proof checkers for each theory
+  if (d_pnm)
+  {
+    ProofChecker* pc = d_pnm->getChecker();
+    for (theory::TheoryId id = theory::THEORY_FIRST; id < theory::THEORY_LAST;
```
Comment (ajreynol) [inline src/smt/smt_solver.cpp:73]: I'd prefer this loop was inside of TheoryEngine::initializeProofChecker, meaning TheoryEngine::initializeProofChecker should not take a TheoryId.

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/smt/smt_solver.cpp (hunk 1 above)
Comment (ajreynol) [reply src/smt/smt_solver.cpp:73]: I don't think there is a reason to be this fine-grained.  I think it'd best to keep it simple, e.g. say a special preprocessing pass used an arithmetic rule.

Comment (HanielB) [reply src/smt/smt_solver.cpp:73]: Ah, so you mean we should always just initialize them all?

Comment (ajreynol) [reply src/smt/smt_solver.cpp:73]: Yes

Comment (HanielB) [reply src/smt/smt_solver.cpp:73]: I see. I'll update.

File: src/theory/bags/theory_bags.h [hunk 2]
```diff
@@ -28,6 +28,9 @@
 #include "theory/theory_eq_notify.h"
 
 namespace cvc5 {
+
+class ProofRuleChecker;
```
Comment (ajreynol) [inline src/theory/bags/theory_bags.h:32]: This shouldn't be necessary.

File: src/theory/fp/theory_fp.h [hunk 3]
```diff
@@ -32,6 +32,9 @@
 #include "theory/uf/equality_engine.h"
 
 namespace cvc5 {
+
+class ProofRuleChecker;
```
Comment (ajreynol) [inline src/theory/fp/theory_fp.h:36]: Ditto.

File: src/theory/sep/theory_sep.h [hunk 4]
```diff
@@ -32,6 +32,9 @@
 #include "util/statistics_registry.h"
 
 namespace cvc5 {
+
+class ProofRuleChecker;
```
Comment (ajreynol) [inline src/theory/sep/theory_sep.h:36]: Ditto.

File: src/theory/sets/theory_sets.h [hunk 5]
```diff
@@ -30,6 +30,9 @@
 #include "theory/uf/equality_engine.h"
 
 namespace cvc5 {
+
+class ProofRuleChecker;
```
Comment (ajreynol) [inline src/theory/sets/theory_sets.h:34]: Ditto.

**Review by ajreynol — CHANGES_REQUESTED** [review]
A few more minor requests.

**Review by ajreynol — APPROVED** [review]
LGTM

**Pull request merged**